sdk/go/pulumi/types.go

```diff
@@ -603,8 +603,12 @@ func awaitInputs(ctx context.Context, v, resolved reflect.Value) (bool, bool, er
 		}
 	case reflect.Ptr:
 		if !v.IsNil() {
-			resolved.Set(reflect.New(resolved.Type().Elem()))
-			return awaitInputs(ctx, v.Elem(), resolved.Elem())
+			if resolved.Kind() == reflect.Ptr {
+				resolved.Set(reflect.New(resolved.Type().Elem()))
+				return awaitInputs(ctx, v.Elem(), resolved.Elem())
+			}
+			// resolved is a struct
```

Comment:
I'm surprised that this is possible: I'd have thought that the assert on line 589 (https://github.com/pulumi/pulumi/pull/3986/files#diff-b79899f0b74b310adc08a01fbcb77e8eR589) would have failed in this case. Do you know why it doesn't?

Reply:
The error isn't because they aren't assignable. The panic is due to `resolved.Type().Elem()` which panics if the kind is not one of a specified list:

from libexec/src/reflect/type.go
```go
func (t *rtype) Elem() Type {
	switch t.Kind() {
	case Array:
		tt := (*arrayType)(unsafe.Pointer(t))
		return toType(tt.elem)
	case Chan:
		tt := (*chanType)(unsafe.Pointer(t))
		return toType(tt.elem)
	case Map:
		tt := (*mapType)(unsafe.Pointer(t))
		return toType(tt.elem)
	case Ptr:
		tt := (*ptrType)(unsafe.Pointer(t))
		return toType(tt.elem)
	case Slice:
		tt := (*sliceType)(unsafe.Pointer(t))
		return toType(tt.elem)
	}
	panic("reflect: Elem of invalid type")
}
```

Reply:
> The error isn't because they aren't assignable.

A pointer value is not assignable to a struct value, which is why I would have thought that assertion would fire.

Reply:
Ah, yes. The original panic was before the assertion:

https://github.com/pulumi/pulumi/pull/3986/files#diff-b79899f0b74b310adc08a01fbcb77e8eR581